packages/heading/Heading.tsx

```diff
@@ -1,34 +1,65 @@
-import React, { ForwardedRef, forwardRef } from 'react'
-import { HeadingStyle, H1Style, H2Style, H3Style } from './HeadingStyles'
-import { HeadingProps, HProps } from './types'
+import { ForwardedRef, forwardRef } from 'react'
+import {
+  HeadingStyle,
+  H1Style,
+  H2Style,
+  H3Style,
+  H1StyleProps,
+  H2StyleProps,
+  H3StyleProps,
+} from './HeadingStyles'
+import { InferStyledComponentPropsWithoutRef } from '../utils'
+import { HeadingColors, HeadingSizes } from './constants'
 
-function Heading(props: HeadingProps, ref?: ForwardedRef<HTMLDivElement>) {
-  const { size = 'md', color = 'text', ...rest } = props
-  return <HeadingStyle $size={size} $color={color} ref={ref} {...rest} />
+export type HeadingProps = InferStyledComponentPropsWithoutRef<
+  typeof HeadingStyle
+> & {
+  color?: HeadingColors
+  size?: HeadingSizes
+}
+function Heading(
+  { size = 'md', color = 'text', ...rest }: HeadingProps,
+  ref?: ForwardedRef<HTMLDivElement>
+) {
+  return <HeadingStyle {...rest} $size={size} $color={color} ref={ref} />
 }
-
 export default forwardRef(Heading)
 
+export type HProps<T extends 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'> =
+  T extends 'h1'
+    ? H1Props
```

Comment:
please check but it should probably work. also, please either remove h4-h6 or add their support

```suggestion
type HeaderKinds = {
  h1: H1Props,
  h2: H2Props,
  h3: H3Props
}
export type HProps<T extends keyof HeaderKinds> =
  keyof HeaderKinds extends T ? never : HeaderKinds[T]
```

Reply:
also, props are probably the same, better to use some more generalized type

Reply:
This was made for backward compatibility, I didn't plan major change, will consider removing it since it's not actually needed in ui kit